### ISSUE TITLE
DL3059 does not flag Consecutive Commands when chaining identical commands

### ISSUE THREAD

**Issue by mmerbes** [hadolint/hadolint#1207]
I understand that this is a controversial rule but I have a use case where commands should be flagged but are not. 

```
RUN apt-get -qqy update && apt-get -qqy --no-install-recommends install libfontconfig
RUN apt-get -qqy update && apt-get -qqy --no-install-recommends install ttf-ubuntu-font-family
RUN apt-get -qqy update && apt-get -qqy --no-install-recommends install fonts-noto-color-emoji
```
Does correctly get flagged for failing DL3059, but something like

```
RUN apt-get -qqy update && apt-get -qqy --no-install-recommends install libfontconfig && rm -rf /var/lib/apt/lists/* && apt-get -qqy clean
RUN apt-get -qqy update && apt-get -qqy --no-install-recommends install ttf-ubuntu-font-family && rm -rf /var/lib/apt/lists/* && apt-get -qqy clean
RUN apt-get -qqy update && apt-get -qqy --no-install-recommends install fonts-noto-color-emoji && rm -rf /var/lib/apt/lists/* && apt-get -qqy clean
```
This however does not despite being basically identical commands with only the packages being different.  

Both these cases should be flagged if the chained commands are identical.

**Comment by m-ildefons** [comment]
Hi,

this is not a bug, it's a deliberate design.

The rule DL3059 (Multiple consecutive `RUN` instructions. Consider consolidation.) only triggers if the commands within the RUN instructions do not chain together with `&&`.

The thought behind this is, if you do something like
```
RUN foo
RUN bar
```
it's two instructions that may well be consolidated to reduce the number of layers.
However if you do
```
RUN foo && bar
RUN baz
```
it's clear that the user knows how to consolidate commands into a single RUN instruction and they deliberately placed two separate RUN instructions (e.g. to improve caching behavior for building).
Therefore the rule only triggers if you don't already chain commands together with `&&`.

In your cases, obviously each RUN instruction contains multiple commands chained together, so it is assumed that you deliberately placed them into separate RUN instructions.

- - -

Hadolint needs to strike a balance and the rules need to catch simple things a beginner might not think about but not get too much in the way for advanced users who know very well what they are doing.
This rule tries not to be too invasive for advanced users who deliberately break their Dockerfiles down to get the best possible caching behavior. In the process it will not catch all cases where RUN instructions can be consolidated, because it assumes that the user may have places multiple consecutive instructions on purpose.
